File: tests/Background.py
#!/usr/bin/env python
"""
Tests for Background

Run with:
   ./Background.py
or
   python
   >>> import Background; Background.run()
"""

import math
import os
import pdb  # we may want to say pdb.set_trace()
import sys
import unittest

import lsst.utils.tests as utilsTests
import lsst.pex.exceptions
import lsst.afw.image.imageLib as afwImage
import lsst.afw.math as afwMath
import lsst.afw.display.ds9 as ds9
import eups

try:
    type(display)
except NameError:
    display = False


# ==== summary to currently implemented tests ====
# getPixel: tests basic functionality of getPixel() method (floats)
# BackgroundTestImages: tests Laher's afwdata/Statistics/*.fits images (doubles)
# testRamp: make sure a constant slope is *exactly* reproduced by the spline model
# testParabola: make sure a quadratic map is *well* reproduced by the spline model

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class BackgroundTestCase(unittest.TestCase):
    
    """A test case for Background"""
    def setUp(self):
        self.val = 10
        self.image = afwImage.ImageF(100, 200); self.image.set(self.val)

    def tearDown(self):
        del self.image


        #self.assertAlmostEqual(mean[1], sd/math.sqrt(image2.getWidth()*image2.getHeight()), 10)

    def testgetPixel(self):
	"""Test the getPixel() function"""


	xcen, ycen = 50, 100
	bgCtrl = afwMath.BackgroundControl(afwMath.NATURAL_SPLINE)
	bgCtrl.setNxSample(3)
	bgCtrl.setNySample(3)
	bgCtrl.sctrl.setNumIter(3)
	bgCtrl.sctrl.setNumSigmaClip(3)
	back = afwMath.BackgroundF(self.image, bgCtrl)
	mid = back.getPixel(xcen,ycen)
	
        self.assertEqual(back.getPixel(xcen,ycen), self.val)


    def testBackgroundTestImages(self):

	imginfolist = []
        imginfolist.append( ["v1_i1_g_m400_s20_f.fits", 400.05551471441612] ) # cooked to known value
        #imgfiles.append("v1_i1_g_m400_s20_u16.fits")
        #imgfiles.append("v1_i2_g_m400_s20_f.fits"
        #imgfiles.append("v1_i2_g_m400_s20_u16.fits")
        #imgfiles.append("v2_i1_p_m9_f.fits")
        #imgfiles.append("v2_i1_p_m9_u16.fits")
        #imgfiles.append("v2_i2_p_m9_f.fits")
        #imgfiles.append("v2_i2_p_m9_u16.fits")
        
        afwdata_dir = eups.productDir("afwdata")
	for imginfo in imginfolist:

	    imgfile, center_value = imginfo

            img_path = afwdata_dir + "/Statistics/" + imgfile;

            # get the image and header
            dimg = afwImage.DecoratedImageD(img_path)
	    img = dimg.getImage()
            fitsHdr = dimg.getMetadata(); # the FITS header

            # get the True values of the mean and stdev
            req_mean  = fitsHdr.getAsDouble("MEANREQ")
            req_stdev = fitsHdr.getAsDouble("SIGREQ")
            naxis1 = img.getWidth()
            naxis2 = img.getHeight()
            
            # create a background control object
            bctrl = afwMath.BackgroundControl(afwMath.NATURAL_SPLINE);
            bctrl.setNxSample(3);
            bctrl.setNySample(3);
	    
            # run the background constructor and call the getPixel() and getImage() functions.
            backobj = afwMath.BackgroundD(img,bctrl)

	    pix_per_subimage = img.getWidth()*img.getHeight()/(bctrl.getNxSample()*bctrl.getNySample())
	    stdev_interp = req_stdev/math.sqrt(pix_per_subimage)
	    
            # test getPixel()
            testval = backobj.getPixel(naxis1/2, naxis2/2)
	    self.assertEqual( testval, center_value )
            self.assertTrue( abs(testval - req_mean) < 2*stdev_interp )

            # test getImage() by checking the center pixel
            bimg = backobj.getImageD()
            testImgval = bimg.get(naxis1/2,naxis2/2)
            self.assertTrue( abs(testImgval - req_mean) < 2*stdev_interp )
	    

    def testRamp(self):

	# make a ramping image (spline should be exact for linear increasing image
	nx = 512
	ny = 512
	rampimg = afwImage.ImageD(nx,ny)
	dzdx, dzdy, z0 = 0.1, 0.2, 10000.0

	for x in range(nx):
	    for y in range(ny):
		rampimg.set(x, y, dzdx*x + dzdy*y + z0)
	
	# check corner, edge, and center pixels
	bctrl = afwMath.BackgroundControl(afwMath.NATURAL_SPLINE);
	bctrl.setNxSample(6);
	bctrl.setNySample(6);
	bctrl.sctrl.setNumSigmaClip(20.0)  # something large enough to avoid clipping entirely
	bctrl.sctrl.setNumIter(1)
	backobj = afwMath.BackgroundD(rampimg,bctrl)

	xpixels = [0,nx/2,nx-1]
	ypixels = [0,ny/2,ny-1]
	for xpix in xpixels:
	    for ypix in ypixels:
		testval = backobj.getPixel(xpix, ypix)
		self.assertAlmostEqual( testval, rampimg.get(xpix,ypix),10 )
		

    def testParabola(self):

	# make an image which varies parabolicly (spline should be exact for 2rd order polynomial)
	nx = 512
	ny = 512
	parabimg = afwImage.ImageD(nx,ny)
	d2zdx2, d2zdy2, dzdx, dzdy, z0 = -1.0e-4, -1.0e-4, 0.1, 0.2, 10000.0  # no cross-terms

	for x in range(nx):
	    for y in range(ny):
		parabimg.set(x, y, d2zdx2*x*x + d2zdy2*y*y + dzdx*x + dzdy*y + z0)
	
	# check corner, edge, and center pixels
	bctrl = afwMath.BackgroundControl(afwMath.NATURAL_SPLINE);
	bctrl.setNxSample(16);
	bctrl.setNySample(16);
	bctrl.sctrl.setNumSigmaClip(10.0)  
	bctrl.sctrl.setNumIter(1)
	backobj = afwMath.BackgroundD(parabimg,bctrl)

	# debug
	#bimg = backobj.getImageD()
	#ds9.mtv(parabimg)
	#ds9.mtv(bimg, frame=1)
	#parabimg.writeFits("a.fits")
	#bimg.writeFits("b.fits")
	
	xpixels = [0,nx/2,nx-1]
	ypixels = [0,ny/2,ny-1]
	for xpix in xpixels:
	    for ypix in ypixels:
		testval = backobj.getPixel(xpix, ypix)
		realval = parabimg.get(xpix,ypix)
		#print xpix, ypix, testval, realval
		# quadratic terms skew the averages of the subimages and the clipped mean for
		# a subimage != value of center pixel.  1/20 counts on a 10000 count sky
		#  is a fair (if arbitrary) test.
		self.assertTrue( (testval - realval) < 0.05 )

    def testCFHT(self):
        """Test background subtraction on some real CFHT data"""

        mi = afwImage.MaskedImageF(os.path.join(eups.productDir("afwdata"), "CFHT", "D4", "cal-53535-i-797722_1"))
        mi = mi.Factory(mi, afwImage.BBox(afwImage.PointI(32, 2), afwImage.PointI(2079, 4609)))
        mi.setXY0(afwImage.PointI(0, 0))
        
	bctrl = afwMath.BackgroundControl(afwMath.NATURAL_SPLINE);
	bctrl.setNxSample(16);
	bctrl.setNySample(16);
	bctrl.sctrl.setNumSigmaClip(3.0)  
	bctrl.sctrl.setNumIter(2)
	backobj = afwMath.BackgroundF(mi.getImage(), bctrl)

        if display:
            ds9.mtv(mi, frame=0)

        im = mi.getImage(); im -= backobj.getImageF()

        if display:
            ds9.mtv(mi, frame=1)

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""

    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(BackgroundTestCase)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)
    return unittest.TestSuite(suites)

def run(exit=False):
    """Run the tests"""
    utilsTests.run(suite(), exit)

if __name__ == "__main__":
    run(True)

File: tests/Background.cc
// -*- LSST-C++ -*-
#include <iostream>
#include <cmath>
#include <vector>

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE Background

#include "boost/test/unit_test.hpp"
#include "boost/test/floating_point_comparison.hpp"

#include "lsst/afw/image/Image.h"
#include "lsst/afw/math/Interpolate.h"
#include "lsst/afw/math/Background.h"

using namespace std;
namespace image = lsst::afw::image;
namespace math = lsst::afw::math;

typedef image::Image<float> Image;
typedef image::DecoratedImage<float> DecoratedImage;

BOOST_AUTO_TEST_CASE(BackgroundBasic) {

    int nX = 40;
    int nY = 40;
    Image img(nX, nY);
    Image::Pixel const pixVal = 10000;
    img = pixVal;

    {
        int xcen = nX/2;
        int ycen = nY/2;
        math::BackgroundControl bgCtrl("AKIMA_SPLINE");
        // test methods native BackgroundControl
        bgCtrl.setNxSample(5);
        bgCtrl.setNySample(5);
        // test methods for public stats objects in bgCtrl
        bgCtrl.sctrl.setNumSigmaClip(3);
        bgCtrl.sctrl.setNumIter(3);
        math::Background back = math::makeBackground(img, bgCtrl);
        double const TESTVAL = back.getPixel(xcen, ycen);
        
        image::Image<float>::Ptr bImage = back.getImage<float>();
        Image::Pixel const testFromImage = *(bImage->xy_at(xcen, ycen));
        
        BOOST_CHECK_EQUAL(TESTVAL, pixVal);
        BOOST_CHECK_EQUAL(TESTVAL, testFromImage);

    }

}

BOOST_AUTO_TEST_CASE(BackgroundTestImages) {

    {
        vector<string> imgfiles;
        imgfiles.push_back("v1_i1_g_m400_s20_f.fits");
        imgfiles.push_back("v1_i1_g_m400_s20_u16.fits");
        //imgfiles.push_back("v1_i2_g_m400_s20_f.fits");
        //imgfiles.push_back("v1_i2_g_m400_s20_u16.fits");
        //imgfiles.push_back("v2_i1_p_m9_f.fits");
        //imgfiles.push_back("v2_i1_p_m9_u16.fits");
        //imgfiles.push_back("v2_i2_p_m9_f.fits");
        //imgfiles.push_back("v2_i2_p_m9_u16.fits");
        
        string afwdata_dir = getenv("AFWDATA_DIR");
        for (vector<string>::iterator imgfile = imgfiles.begin(); imgfile != imgfiles.end(); ++imgfile) {
            
            string img_path = afwdata_dir + "/Statistics/" + *imgfile;

            // get the image and header
            DecoratedImage dimg(img_path);
            Image::Ptr img = dimg.getImage();
            lsst::daf::base::PropertySet::Ptr fitsHdr = dimg.getMetadata(); // the FITS header

            // get the true values of the mean and stdev
            float reqMean = static_cast<float>(fitsHdr->getAsDouble("MEANREQ"));
            float reqStdev = static_cast<float>(fitsHdr->getAsDouble("SIGREQ"));

            int const width = img->getWidth();
            int const height = img->getHeight();
            
            // create a background control object
            math::BackgroundControl bctrl(math::Interpolate::AKIMA_SPLINE);
            bctrl.setNxSample(5);
            bctrl.setNySample(5);
            float stdevSubimg = reqStdev / sqrt(width*height/(bctrl.getNxSample()*bctrl.getNySample()));

            // run the background constructor and call the getPixel() and getImage() functions.
            math::Background backobj = math::makeBackground(*img, bctrl);

            // test getPixel()
            float testval = static_cast<float>(backobj.getPixel(width/2, height/2));
            BOOST_REQUIRE( fabs(testval - reqMean) < 2.0*stdevSubimg );

            // test getImage() by checking the center pixel
            image::Image<float>::Ptr bimg = backobj.getImage<float>();
            float testImgval = static_cast<float>(*(bimg->xy_at(width/2, height/2)));
            BOOST_REQUIRE( fabs(testImgval - reqMean) < 2.0*stdevSubimg );
            
        }
    }
        
}



BOOST_AUTO_TEST_CASE(BackgroundRamp) {

    {
        
        // make a ramping image (spline should be exact for linear increasing image
        int const nX = 512;
        int const nY = 512;
        image::Image<double> rampimg = image::Image<double>(nX, nY);
        double dzdx = 0.1;
        double dzdy = 0.2;
        double z0 = 10000.0;

        for (int i = 0; i < nX; ++i) {
            double x = static_cast<double>(i);
            for ( int j = 0; j < nY; ++j) {
                double y = static_cast<double>(j);
                *rampimg.xy_at(i, j) = dzdx*x + dzdy*y + z0;
            }
        }
        
        // check corner, edge, and center pixels
        math::BackgroundControl bctrl = math::BackgroundControl(math::Interpolate::AKIMA_SPLINE);
        bctrl.setNxSample(6);
        bctrl.setNySample(6);
        bctrl.sctrl.setNumSigmaClip(20.0);  // something large enough to avoid clipping entirely
        bctrl.sctrl.setNumIter(1);
        math::Background backobj = math::Background(rampimg, bctrl);

        // test the values at the corners and in the middle
        int ntest = 3;
        for (int i = 0; i < ntest; ++i) {
            int xpix = i*(nX - 1)/(ntest - 1);
            for (int j = 0; j < ntest; ++j) {
                int ypix = j*(nY - 1)/(ntest - 1);
                double testval = backobj.getPixel(xpix, ypix);
                double realval = *rampimg.xy_at(xpix, ypix);
                BOOST_CHECK_CLOSE( testval, realval, 1.0e-10 );
            }
        }
                    
    }

}
BOOST_AUTO_TEST_CASE(BackgroundParabola) {

    {
        
        // make an image which varies parabolicly (spline should be exact for 2rd order polynomial)
        int const nX = 512;
        int const nY = 512;
        image::Image<double> parabimg = image::Image<double>(nX, nY);
        double d2zdx2 = -1.0e-4;
        double d2zdy2 = -1.0e-4;
        double dzdx   = 0.1;
        double dzdy   = 0.2;
        double z0 = 10000.0;  // no cross-terms

        for ( int i = 0; i < nX; ++i ) {
            for ( int j = 0; j < nY; ++j ) {
                *parabimg.xy_at(i, j) = d2zdx2*i*i + d2zdy2*j*j + dzdx*i + dzdy*j + z0;
            }
        }
        
        // check corner, edge, and center pixels
        math::BackgroundControl bctrl = math::BackgroundControl(math::Interpolate::CUBIC_SPLINE);
        bctrl.setNxSample(24);
        bctrl.setNySample(24);
        bctrl.sctrl.setNumSigmaClip(10.0);
        bctrl.sctrl.setNumIter(1);
        math::Background backobj = math::Background(parabimg, bctrl);

        // debug
        //bimg = backobj.getImageD()
        //ds9.mtv(parabimg)
        //ds9.mtv(bimg, frame=1)
        //parabimg.writeFits("a.fits")
        //bimg.writeFits("b.fits")

        // check the values at the corners and int he middle
        int const ntest = 3;
        for (int i = 0; i < ntest; ++i) {
            int xpix = i*(nX - 1)/(ntest - 1);
            for (int j = 0; j < ntest; ++j) {
                int ypix = j*(nY - 1)/(ntest - 1);
                double testval = backobj.getPixel(xpix, ypix);
                double realval = *parabimg.xy_at(xpix, ypix);
                //print xpix, ypix, testval, realval
                // quadratic terms skew the averages of the subimages and the clipped mean for
                // a subimage != value of center pixel.  1/20 counts on a 10000 count sky
                //  is a fair (if arbitrary) test.
                BOOST_CHECK_CLOSE( testval, realval, 0.05 );
            }
        }
    }
}

